src/persist/collect-idb-databases.ts

```diff
@@ -0,0 +1,162 @@
+import {ClientMap, getClients} from './clients.js';
+import {assertNotTempHash} from '../hash.js';
+import {dropStore} from '../kv/idb-store.js';
+import type {IDBDatabasesStore, IndexedDBDatabase} from './idb-databases-store';
+import {initBgIntervalProcess} from './bg-interval.js';
+import type {LogContext} from '@rocicorp/logger';
+import {sleep} from '../sleep.js';
+import {AbortError} from '../abort-error.js';
+import {REPLICACHE_FORMAT_VERSION} from '../replicache.js';
+
+// How frequently to try to collect
+const COLLECT_INTERVAL_MS = 12 * 60 * 60 * 1000; // 12 hours
+
+// If an IDB database is older than MAX_AGE, then it can be collected.
+const MAX_AGE = 3 * 30 * 24 * 60 * 60 * 1000; // 3 months
+
+// We delay the initial collection to prevent doing it at startup.
+const COLLECT_DELAY = 5 * 60 * 1000; // 5 minutes
+
+export function initCollectIDBDatabases(
+  idbDatabasesStore: IDBDatabasesStore,
+  lc: LogContext,
+  signal: AbortSignal,
+): void {
+  void sleepFiveAndCollect(idbDatabasesStore, signal);
+
+  const stopInterval = initBgIntervalProcess(
+    'CollectIDBDatabases',
+    async () => {
+      await collectIDBDatabases(
+        idbDatabasesStore,
+        signal,
+        Date.now(),
+        MAX_AGE,
+        undefined,
+      );
+    },
+    COLLECT_INTERVAL_MS,
+    lc,
+  );
+
+  signal.addEventListener('abort', stopInterval);
+}
+
+async function sleepFiveAndCollect(
+  idbDatabasesStore: IDBDatabasesStore,
+  signal: AbortSignal,
+) {
+  try {
+    await sleep(COLLECT_DELAY, signal);
+
+    await collectIDBDatabases(
+      idbDatabasesStore,
+      signal,
+      Date.now(),
+      MAX_AGE,
+      defaultNewDagStore,
+    );
+  } catch (e) {
+    if (e instanceof AbortError) {
+      return;
+    }
+    throw e;
+  }
+}
+
+export async function collectIDBDatabases(
+  idbDatabasesStore: IDBDatabasesStore,
+  signal: AbortSignal,
+  now: number,
+  maxAge: number,
+  newDagStore = defaultNewDagStore,
+): Promise<void> {
+  const databases = await idbDatabasesStore.getDatabases();
+
+  const dbs = Object.values(databases) as IndexedDBDatabase[];
+  const canCollectResults = await Promise.all(
+    dbs.map(
+      async db =>
+        [
+          db.name,
+          await canCollectDatabase(db, now, maxAge, newDagStore),
+        ] as const,
+    ),
+  );
+
+  const namesToRemove = canCollectResults
+    .filter(result => result[1])
+    .map(result => result[0]);
+
+  // Try to remove the databases in parallel. Don't let a single reject fail the
+  // other ones. We will check for failures afterwards.
+  const dropStoreResults = await Promise.allSettled(
+    namesToRemove.map(async name => {
+      await dropStore(name);
+      return name;
+    }),
+  );
+
+  const idbRemovedNames: string[] = [];
+  const errors: unknown[] = [];
+  for (const result of dropStoreResults) {
+    if (result.status === 'fulfilled') {
+      idbRemovedNames.push(result.value);
+    } else {
+      errors.push(result.reason);
```

Comment:
Is there a reasonable way for us to detect if the delete failed because the db no longer exists, so we can remove that db from the meta table?

Reply:
`indexedDB.deleteDatabase('no-such-db')` succeeds.

Reply:
The failure case I found is that the DB is in use.